### ISSUE TITLE
Dioxus router removes hash segments

### ISSUE THREAD

**Issue by ealmloff** [DioxusLabs/dioxus#1994]
**Problem**

The dioxus router removes any hash segments when parsing the route.

**Steps To Reproduce**

- Navigate to https://dioxuslabs.com/learn/0.4/getting_started#setup-guides
- See the url change to https://dioxuslabs.com/learn/0.4/getting_started

**Expected behavior**

The hash segment should remain

**Environment:**
 - Dioxus version: `master`
 - Rust version: `nightly`
 - OS info: Linux
 - App platform: `web`

**Questionnaire**

- [ ] I'm interested in fixing this myself but don't know where to start
- [ ] I would like to fix and I have a solution
- [ ] I don't have time to fix this right now, but maybe later

**Comment by SET001** [comment]
URL fragments might have `query-string` format.

here is an example URL from [google oauth docs](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#oauth-2.0-endpoints_1) -`https://oauth2.example.com/callback#access_token=4/P7q7W91&token_type=Bearer&expires_in=3600` How this should be parsed?

I image route definition similar to this:
```
#[route("/callback#:fragment")]
```
and then `fragment` should be something that implements `FromQuery`? Right?

**Comment by ealmloff** [comment]
> URL fragments might have `query-string` format.
> 
> here is an example URL from [google oauth docs](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow#oauth-2.0-endpoints_1) -`https://oauth2.example.com/callback#access_token=4/P7q7W91&token_type=Bearer&expires_in=3600` How this should be parsed?
> 
> I image route definition similar to this:
> 
> ```
> #[route("/callback#:fragment")]
> ```
> 
> and then `fragment` should be something that implements `FromQuery`? Right?

That route definition looks correct. Fragments may look like a query, but we don't want to force them to look like a traditional query segment (`name=bla&surname=bla`). We can use a trait with an identical signature to `FromQuery` called `FromFragment` or just accept `T: FromStr<Err = core::convert::Infallible>` instead of `FromQuery` and `FromFragment`

**Comment by kalzoo** [comment]
Just a note: this bug prevents the use of the OAuth2 Implicit Grant flow with some providers. That includes Google as @SET001 mentioned on discord, and also AWS: https://docs.aws.amazon.com/cognito/latest/developerguide/authorization-endpoint.html. 

Implicit grant flow returns tokens in the fragment; auth code flow returns them in the query string. I tried this workaround in the hope that the first render would take place before the fragment was removed by Dioxus, but no luck:

```rust
    use web_sys::window;
    let hash = window().unwrap().location().hash().unwrap();
    let params_parsed: HashMap<String, String> = serde_urlencoded::from_str(&hash).unwrap();
```
